File: codifide/parser/expr_parser.py
"""Recursive-descent parser for Codifide expressions.

Grammar (v0):

    expr    := bind | concat
    bind    := ident "<-" concat ";" expr
    concat  := atom ("++" atom)*
    atom    := number
             | string
             | "bottom"
             | ident "(" args ")"        -- call
             | ident                     -- ref / dotted ref
             | "(" expr ")"
    args    := expr ("," expr)*

Comparisons and booleans used in pre/post/guard expressions are surfaced as
calls to primitive names: `lt(a, b)`, `ge(a, b)`, `and(...)`, etc. The surface
layer therefore supports `<`, `<=`, `>`, `>=`, `==`, `!=` as infix sugar that
desugars to those primitives. This keeps the core AST tiny.

Believe-dispatch and sequences are handled by the line-oriented outer parser,
not here. This module parses a single expression fragment.
"""
from __future__ import annotations

from typing import List, Optional

from ..core.types import (
    Attr,
    BottomExpr,
    Call,
    Concat,
    Expr,
    If,
    Lit,
    Ref,
)
from .lexer import Token, lex_expr


class ExprParseError(Exception):
    pass


# Infix comparison operators that desugar to primitive calls. Ordered by
# length so `<=` is matched before `<`.
_INFIX = [
    ("<=", "le"),
    (">=", "ge"),
    ("==", "eq"),
    ("!=", "ne"),
    ("<",  "lt"),
    (">",  "gt"),
    ("and", "and"),
    ("or",  "or"),
]


def parse_expr(src: str) -> Expr:
    """Parse a single expression fragment to a canonical Expr."""
    # Rewrite infix comparisons to function-call syntax before lexing. This is
    # a cheap way to keep the lexer and parser tiny while still offering a
    # familiar surface for humans bootstrapping tests.
    rewritten = _desugar_infix(src)
    tokens = lex_expr(rewritten)
    parser = _Parser(tokens)
    result = parser.parse_expr()
    if not parser.eof():
        raise ExprParseError(f"trailing tokens after expression: {parser.peek()}")
    return result


def _desugar_infix(src: str) -> str:
    # We only desugar top-level occurrences outside strings. Because the
    # surface is simple, a character scan suffices.
    out: List[str] = []
    i = 0
    n = len(src)
    in_str = False
    while i < n:
        c = src[i]
        if c == '"':
            in_str = not in_str
            out.append(c)
            i += 1
            continue
        if in_str:
            out.append(c)
            i += 1
            continue
        matched = False
        for op, name in _INFIX:
            if src.startswith(op, i):
                # Word operators require boundaries; symbolic ones do not.
                if op.isalpha():
                    # ``_`` is part of an identifier as far as Codifide
                    # names are concerned. Without treating it as
                    # alphanumeric, the substring ``or`` inside
                    # ``greet_or_refuse`` passes the boundary check
                    # (``_`` is not alphanumeric in Python's sense)
                    # and the desugarer splits the identifier. Same
                    # class of bug as the ``and(...)`` case below;
                    # both caught while writing the 2026-05-11
                    # assessment battery.
                    def _is_ident_char(ch: str) -> bool:
                        return ch.isalnum() or ch == "_"
                    left_ok = i == 0 or not _is_ident_char(src[i - 1])
                    right_ok = (
                        i + len(op) >= n
                        or not _is_ident_char(src[i + len(op)])
                    )
                    if not (left_ok and right_ok):
                        continue
                    # If the word operator is followed (after optional
                    # whitespace) by '(', it's a function call, not
                    # infix. Example: `and(p, q)` is a call to the
                    # ``and`` primitive with two arguments, NOT an
                    # infix expression. Without this guard the
                    # desugarer tried to extract left/right operands
                    # from `and(...)` and choked — found while writing
                    # the 2026-05-11 assessment programs.
                    k = i + len(op)
                    while k < n and src[k].isspace():
                        k += 1
                    if k < n and src[k] == "(":
                        continue
                # Rewrite `a OP b` as `name(a, b)`. We find the left operand by
                # walking back through whitespace and a balanced parenthesis
                # or identifier; and the right operand by a lazy bracket scan.
                left, out = _pop_operand(out)
                right, j = _take_operand(src, i + len(op))
                out.append(f"{name}({left}, {right})")
                i = j
                matched = True
                break
        if not matched:
            out.append(c)
            i += 1
    return "".join(out)


def _pop_operand(buf: List[str]) -> (str, List[str]):
    # Collect the trailing operand from the output buffer. We walk back until
    # we've captured a balanced parenthesis group, or an identifier/number.
    s = "".join(buf).rstrip()
    if not s:
        return "", []
    if s.endswith(")"):
        depth = 0
        i = len(s) - 1
        while i >= 0:
            if s[i] == ")":
                depth += 1
            elif s[i] == "(":
                depth -= 1
                if depth == 0:
                    # Include any leading callable identifier.
                    j = i - 1
                    while j >= 0 and (s[j].isalnum() or s[j] in "._"):
                        j -= 1
                    operand = s[j + 1:]
                    return operand, list(s[: j + 1])
            i -= 1
        raise ExprParseError("unbalanced parentheses in infix desugaring")
    # Identifier/number run.
    j = len(s) - 1
    while j >= 0 and (s[j].isalnum() or s[j] in "._"):
        j -= 1
    return s[j + 1:], list(s[: j + 1])


def _take_operand(src: str, start: int) -> (str, int):
    i = start
    n = len(src)
    while i < n and src[i].isspace():
        i += 1
    if i >= n:
        raise ExprParseError("missing right operand")
    # Balanced paren group.
    if src[i] == "(":
        depth = 1
        j = i + 1
        while j < n and depth > 0:
            if src[j] == "(":
                depth += 1
            elif src[j] == ")":
                depth -= 1
            j += 1
        return src[i:j], j
    # Identifier or call.
    j = i
    while j < n and (src[j].isalnum() or src[j] in "._"):
        j += 1
    if j < n and src[j] == "(":
        depth = 1
        j += 1
        while j < n and depth > 0:
            if src[j] == "(":
                depth += 1
            elif src[j] == ")":
                depth -= 1
            j += 1
    return src[i:j], j


class _Parser:
    # Maximum nesting depth for parenthesized sub-expressions. A
    # recursive-descent parser with no bound blows the Python stack on
    # adversarial input (``((...(1)...))`` with a few hundred parens),
    # leaking RecursionError to the host. The fuzz harness at
    # tests/test_parser_fuzz.py catches this; the bound here keeps
    # pathological input from reaching the Python stack limit.
    MAX_PAREN_DEPTH = 256

    def __init__(self, tokens: List[Token]) -> None:
        self.tokens = tokens
        self.pos = 0
        self._paren_depth = 0

    def eof(self) -> bool:
        return self.pos >= len(self.tokens)

    def peek(self) -> Optional[Token]:
        return None if self.eof() else self.tokens[self.pos]

    def take(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def parse_expr(self) -> Expr:
        return self.parse_concat()

    def parse_concat(self) -> Expr:
        left = self.parse_atom()
        parts: List[Expr] = [left]
        while not self.eof() and self.peek().kind == "op" and self.peek().text in ("++", "⊕"):
            self.take()
            parts.append(self.parse_atom())
        if len(parts) == 1:
            return left
        return Concat(tuple(parts))

    def parse_atom(self) -> Expr:
        tok = self.peek()
        if tok is None:
            raise ExprParseError("unexpected end of expression")
        if tok.kind == "num":
            self.take()
            if "." in tok.text:
                return Lit(float(tok.text), type="Float")
            return Lit(int(tok.text), type="Int")
        if tok.kind == "str":
            self.take()
            return Lit(tok.text, type="String")
        if tok.kind == "ident":
            if tok.text == "bottom":
                self.take()
                # Optional reason string: bottom "reason text"
                nxt = self.peek()
                if nxt is not None and nxt.kind == "str":
                    self.take()
                    return BottomExpr(reason=nxt.text)
                return BottomExpr()
            if tok.text == "true":
                self.take()
                return Lit(True, type="Bool")
            if tok.text == "false":
                self.take()
                return Lit(False, type="Bool")
            if tok.text == "if":
                # Inline conditional: ``if cond then a else b``.
                # Short-circuit, unlike candidate-dispatch guards.
                # Added 2026-05-11 (spec amendment).
                self.take()
                cond = self.parse_expr()
                nxt = self.peek()
                if nxt is None or nxt.kind != "ident" or nxt.text != "then":
                    raise ExprParseError(
                        "expected 'then' after `if <cond>`"
                    )
                self.take()
                then_branch = self.parse_expr()
                nxt = self.peek()
                if nxt is None or nxt.kind != "ident" or nxt.text != "else":
                    raise ExprParseError(
                        "expected 'else' after `if <cond> then <expr>`"
                    )
                self.take()
                else_branch = self.parse_expr()
                return If(cond=cond, then_=then_branch, else_=else_branch)
            self.take()
            # Function call?
            nxt = self.peek()
            if nxt is not None and nxt.kind == "punct" and nxt.text == "(":
                self.take()  # (
                args: List[Expr] = []
                p = self.peek()
                if p is None:
                    raise ExprParseError(
                        f"unexpected end of expression after '{tok.text}('"
                    )
                if not (p.kind == "punct" and p.text == ")"):
                    args.append(self.parse_expr())
                    while True:
                        p = self.peek()
                        if p is None:
                            raise ExprParseError(
                                f"unexpected end of expression inside "
                                f"arguments of '{tok.text}'"
                            )
                        if not (p.kind == "punct" and p.text == ","):
                            break
                        self.take()
                        args.append(self.parse_expr())
                closer = self.peek()
                if closer is None or not (
                    closer.kind == "punct" and closer.text == ")"
                ):
                    raise ExprParseError(
                        f"expected ')' to close call to '{tok.text}'"
                    )
                self.take()  # )
                return self._call_or_attr(tok.text, tuple(args))
            # Reference or dotted reference (treated as an Attr chain rooted
            # at the first segment, so `clock.now.hm` becomes
            # Attr(Attr(Ref("clock"), "now"), "hm")).
            return self._dotted_ref(tok.text)
        if tok.kind == "punct" and tok.text == "(":
            self.take()
            if self._paren_depth >= self.MAX_PAREN_DEPTH:
                raise ExprParseError(
                    f"parenthesis nesting exceeds {self.MAX_PAREN_DEPTH}"
                )
            self._paren_depth += 1
            try:
                inner = self.parse_expr()
            finally:
                self._paren_depth -= 1
            closer = self.peek()
            if closer is None or not (
                closer.kind == "punct" and closer.text == ")"
            ):
                raise ExprParseError("expected ')'")
            self.take()
            return inner
        raise ExprParseError(f"unexpected token {tok}")

    # -- helpers ---------------------------------------------------------

    def _dotted_ref(self, name: str) -> Expr:
        if "." not in name:
            return Ref(name)
        head, *rest = name.split(".")
        node: Expr = Ref(head)
        for seg in rest:
            node = Attr(node, seg)
        return node

    def _call_or_attr(self, name: str, args: tuple) -> Expr:
        # A dotted call like `io.say("...")` is still a single Call node whose
        # `fn` is the dotted name. The interpreter resolves that name against
        # both the module symbols and the primitive registry.
        return Call(name, args)

File: crates/codifide-interpreter/src/parser/expr_parser.rs
//! Recursive-descent expression parser for Codifide.
//! Mirrors `codifide/parser/expr_parser.py`.
//!
//! Grammar:
//!   expr    := concat
//!   concat  := atom ("++" atom)*
//!   atom    := number | string | "bottom" | "true" | "false"
//!            | "if" expr "then" expr "else" expr
//!            | ident "(" args ")"
//!            | ident
//!            | "(" expr ")"

use codifide_canonical::ast::{Expr};
use super::lexer::{lex_expr, LexError, Token, TokenKind};
use super::tokens::INFIX_OPS;

#[derive(Debug)]
pub struct ExprParseError(pub String);

impl std::fmt::Display for ExprParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<LexError> for ExprParseError {
    fn from(e: LexError) -> Self { ExprParseError(e.0) }
}

pub fn parse_expr(src: &str) -> Result<Expr, ExprParseError> {
    let rewritten = desugar_infix(src);
    let tokens = lex_expr(&rewritten)?;
    let mut p = Parser::new(tokens);
    let result = p.parse_expr()?;
    if !p.eof() {
        let tok = p.peek().unwrap();
        return Err(ExprParseError(format!("trailing tokens after expression: {:?}", tok.text)));
    }
    Ok(result)
}

// ---------------------------------------------------------------------------
// Infix desugaring — mirrors Python's _desugar_infix
// ---------------------------------------------------------------------------

fn desugar_infix(src: &str) -> String {
    let chars: Vec<char> = src.chars().collect();
    let n = chars.len();
    let mut out: Vec<char> = Vec::with_capacity(n);
    let mut i = 0usize;
    let mut in_str = false;

    while i < n {
        let c = chars[i];
        if c == '"' {
            in_str = !in_str;
            out.push(c);
            i += 1;
            continue;
        }
        if in_str {
            out.push(c);
            i += 1;
            continue;
        }

        let rest: String = chars[i..].iter().collect();
        let mut matched = false;
        for (op, name) in INFIX_OPS {
            if !rest.starts_with(op) { continue; }
            let op_chars: Vec<char> = op.chars().collect();
            let op_len = op_chars.len();

            // Word operators need boundaries
            if op.chars().all(|c| c.is_alphabetic()) {
                let is_ident = |ch: char| ch.is_alphanumeric() || ch == '_';
                let left_ok = i == 0 || !is_ident(chars[i - 1]);
                let right_i = i + op_len;
                let right_ok = right_i >= n || !is_ident(chars[right_i]);
                if !left_ok || !right_ok { continue; }
                // If followed by '(' it's a function call, not infix
                let mut k = i + op_len;
                while k < n && chars[k].is_whitespace() { k += 1; }
                if k < n && chars[k] == '(' { continue; }
            }

            // Pop left operand from out buffer
            let out_str: String = out.iter().collect();
            let (left, new_out) = pop_operand(&out_str);
            out = new_out.chars().collect();

            // Take right operand from src
            let (right, j) = take_operand(&chars, i + op_len);

            // Emit name(left, right)
            let repl = format!("{}({}, {})", name, left, right);
            out.extend(repl.chars());
            i = j;
            matched = true;
            break;
        }
        if !matched {
            out.push(c);
            i += 1;
        }
    }
    out.iter().collect()
}

fn pop_operand(buf: &str) -> (String, String) {
    let s = buf.trim_end();
    if s.is_empty() { return (String::new(), String::new()); }
    if s.ends_with(')') {
        let bytes: Vec<char> = s.chars().collect();
        let mut depth = 0i32;
        let mut i = bytes.len();
        loop {
            if i == 0 { break; }
            i -= 1;
            match bytes[i] {
                ')' => depth += 1,
                '(' => {
                    depth -= 1;
                    if depth == 0 {
                        // Include leading callable identifier
                        let mut j = i;
                        while j > 0 && (bytes[j-1].is_alphanumeric() || bytes[j-1] == '_' || bytes[j-1] == '.') {
                            j -= 1;
                        }
                        let operand: String = bytes[j..].iter().collect();
                        let prefix: String = bytes[..j].iter().collect();
                        return (operand, prefix);
                    }
                }
                _ => {}
            }
        }
        return (s.to_string(), String::new());
    }
    // Identifier/number run
    let bytes: Vec<char> = s.chars().collect();
    let mut j = bytes.len();
    while j > 0 && (bytes[j-1].is_alphanumeric() || bytes[j-1] == '_' || bytes[j-1] == '.') {
        j -= 1;
    }
    let operand: String = bytes[j..].iter().collect();
    let prefix: String = bytes[..j].iter().collect();
    (operand, prefix)
}

fn take_operand(chars: &[char], start: usize) -> (String, usize) {
    let n = chars.len();
    let mut i = start;
    while i < n && chars[i].is_whitespace() { i += 1; }
    if i >= n { return (String::new(), i); }
    if chars[i] == '(' {
        let mut depth = 1i32;
        let mut j = i + 1;
        while j < n && depth > 0 {
            match chars[j] { '(' => depth += 1, ')' => depth -= 1, _ => {} }
            j += 1;
        }
        return (chars[i..j].iter().collect(), j);
    }
    let mut j = i;
    while j < n && (chars[j].is_alphanumeric() || chars[j] == '_' || chars[j] == '.') { j += 1; }
    if j < n && chars[j] == '(' {
        let mut depth = 1i32;
        j += 1;
        while j < n && depth > 0 {
            match chars[j] { '(' => depth += 1, ')' => depth -= 1, _ => {} }
            j += 1;
        }
    }
    (chars[i..j].iter().collect(), j)
}

// ---------------------------------------------------------------------------
// Recursive-descent parser
// ---------------------------------------------------------------------------

const MAX_PAREN_DEPTH: usize = 256;

struct Parser {
    tokens: Vec<Token>,
    pos: usize,
    paren_depth: usize,
}

impl Parser {
    fn new(tokens: Vec<Token>) -> Self { Parser { tokens, pos: 0, paren_depth: 0 } }
    fn eof(&self) -> bool { self.pos >= self.tokens.len() }
    fn peek(&self) -> Option<&Token> { self.tokens.get(self.pos) }
    fn take(&mut self) -> &Token { let t = &self.tokens[self.pos]; self.pos += 1; t }

    fn parse_expr(&mut self) -> Result<Expr, ExprParseError> {
        self.parse_concat()
    }

    fn parse_concat(&mut self) -> Result<Expr, ExprParseError> {
        let left = self.parse_atom()?;
        let mut parts = vec![left];
        while let Some(tok) = self.peek() {
            if tok.kind == TokenKind::Op && (tok.text == "++" || tok.text == "⊕") {
                self.take();
                parts.push(self.parse_atom()?);
            } else {
                break;
            }
        }
        if parts.len() == 1 { return Ok(parts.remove(0)); }
        Ok(Expr::Concat { parts })
    }

    fn parse_atom(&mut self) -> Result<Expr, ExprParseError> {
        let tok = self.peek().ok_or_else(|| ExprParseError("unexpected end of expression".into()))?;

        match tok.kind {
            TokenKind::Num => {
                let text = self.take().text.clone();
                if text.contains('.') {
                    let v: f64 = text.parse().map_err(|_| ExprParseError(format!("invalid number: {}", text)))?;
                    Ok(Expr::Lit { value: serde_json::json!(v), type_: "Float".into(), conf: 1.0, provenance: "literal".into() })
                } else {
                    let v: i64 = text.parse().map_err(|_| ExprParseError(format!("invalid integer: {}", text)))?;
                    Ok(Expr::Lit { value: serde_json::json!(v), type_: "Int".into(), conf: 1.0, provenance: "literal".into() })
                }
            }
            TokenKind::Str => {
                let text = self.take().text.clone();
                Ok(Expr::Lit { value: serde_json::json!(text), type_: "String".into(), conf: 1.0, provenance: "literal".into() })
            }
            TokenKind::Ident => {
                let text = tok.text.clone();
                match text.as_str() {
                    "bottom" => {
                        self.take();
                        // Optional reason string: bottom "reason text"
                        let reason = if let Some(next) = self.peek() {
                            if next.kind == TokenKind::Str {
                                let r = self.take().text.clone();
                                Some(r)
                            } else {
                                None
                            }
                        } else {
                            None
                        };
                        Ok(Expr::Bottom { reason })
                    }
                    "true"   => { self.take(); Ok(Expr::Lit { value: serde_json::json!(true),  type_: "Bool".into(), conf: 1.0, provenance: "literal".into() }) }
                    "false"  => { self.take(); Ok(Expr::Lit { value: serde_json::json!(false), type_: "Bool".into(), conf: 1.0, provenance: "literal".into() }) }
                    "if" => {
                        self.take();
                        let cond = self.parse_expr()?;
                        // expect "then"
                        match self.peek() {
                            Some(t) if t.kind == TokenKind::Ident && t.text == "then" => { self.take(); }
                            _ => return Err(ExprParseError("expected 'then' after `if <cond>`".into())),
                        }
                        let then_ = self.parse_expr()?;
                        // expect "else"
                        match self.peek() {
                            Some(t) if t.kind == TokenKind::Ident && t.text == "else" => { self.take(); }
                            _ => return Err(ExprParseError("expected 'else' after `if <cond> then <expr>`".into())),
                        }
                        let else_ = self.parse_expr()?;
                        Ok(Expr::If { cond: Box::new(cond), then_: Box::new(then_), else_: Box::new(else_) })
                    }
                    _ => {
                        self.take();
                        // Function call?
                        if let Some(next) = self.peek() {
                            if next.kind == TokenKind::Punct && next.text == "(" {
                                self.take(); // (
                                let mut args = Vec::new();
                                if let Some(p) = self.peek() {
                                    if !(p.kind == TokenKind::Punct && p.text == ")") {
                                        args.push(self.parse_expr()?);
                                        loop {
                                            match self.peek() {
                                                Some(p) if p.kind == TokenKind::Punct && p.text == "," => {
                                                    self.take();
                                                    args.push(self.parse_expr()?);
                                                }
                                                _ => break,
                                            }
                                        }
                                    }
                                }
                                match self.peek() {
                                    Some(p) if p.kind == TokenKind::Punct && p.text == ")" => { self.take(); }
                                    _ => return Err(ExprParseError(format!("expected ')' to close call to '{}'", text))),
                                }
                                return Ok(Expr::Call { fn_: text, args });
                            }
                        }
                        // Dotted ref: clock.now → Attr(Ref("clock"), "now")
                        Ok(dotted_ref(&text))
                    }
                }
            }
            TokenKind::Punct if tok.text == "(" => {
                self.take();
                if self.paren_depth >= MAX_PAREN_DEPTH {
                    return Err(ExprParseError(format!("parenthesis nesting exceeds {}", MAX_PAREN_DEPTH)));
                }
                self.paren_depth += 1;
                let inner = self.parse_expr()?;
                self.paren_depth -= 1;
                match self.peek() {
                    Some(p) if p.kind == TokenKind::Punct && p.text == ")" => { self.take(); }
                    _ => return Err(ExprParseError("expected ')'".into())),
                }
                Ok(inner)
            }
            _ => Err(ExprParseError(format!("unexpected token {:?}", tok.text))),
        }
    }
}

fn dotted_ref(name: &str) -> Expr {
    if !name.contains('.') {
        return Expr::Ref { name: name.to_string() };
    }
    let parts: Vec<&str> = name.splitn(2, '.').collect();
    let head = Expr::Ref { name: parts[0].to_string() };
    // Recursively build Attr chain for multi-segment names
    build_attr(head, parts[1])
}

fn build_attr(target: Expr, rest: &str) -> Expr {
    if !rest.contains('.') {
        return Expr::Attr { target: Box::new(target), name: rest.to_string() };
    }
    let parts: Vec<&str> = rest.splitn(2, '.').collect();
    let mid = Expr::Attr { target: Box::new(target), name: parts[0].to_string() };
    build_attr(mid, parts[1])
}
